chanzuckerberg/czgenepi
# thuang-next-js-138313

### Description

This PR seems like a huge change, but it can be boiled down to the following main points:

1. Replace Create React App related dependencies and scripts, and use Next.js to manage all that for us
2. Create `pages/` directory required by Next.js to host all pages, since they use file system based routing for both static and dynamic routes: https://nextjs.org/docs/routing/introduction
3. Update `docker-compose` to pass `HAPPY_COMMIT` to use as consistent build id for all Next.js servers in a cluster
4. Next.js has its own `next/router`, so we no longer need `react-router`
5. Add `pages/_document.tsx` and `pages/_app.tsx` for hosting global Next.js page configuration
6. Add `src/frontend/next.config.js` to configure Next.js to process `SVG` files as React components, add CSP/HSTS, specify prod build directory (default is `.next/`, but Infra expects `build/`), add `HAPPY_COMMIT` as consistent build id, and make `envVariable.js` available in the browser (see next point)
7. In Next.js, by default `process.env` variables are only available in Node.js and not in the browser, and they allow exposing env variables to browser through prefixing a variable with `NEXT_PUBLIC_`, which I find noisy and affecting across too many stacks and files (Docker files, entrypoint.js, and FE files). So I created `envVariable.js` as the single source of truth for all env variables used in the FE app and export them as process.env variables to the browser, so we don't need `NEXT_PUBLIC_` prefix!
8. Update `package.json` scripts to use Next.js' commands
9. Add [`react-query`](https://react-query.tanstack.com/) for handling API calls, which handles caching, garbage collection, refetch for us automatically
10. Use `import Head from 'next/head'` to add page titles

That's it! PTAL thank you 😄 

#### Issue
https://app.clubhouse.io/genepi/story/138313/migrate-fe-to-next-js

### Test plan

1. Visit: https://thuang-next-1-frontend.dev.genepi.czi.technology
2. The site should be fully functional
3. Some things I explicitly checked are:

1. Aspen logo works in Auth0 - inferred through https://thuang-next-1-frontend.dev.genepi.czi.technology/AspenLogoBlack.svg works
2. All routes work - /samples, /trees, /agreeTerms, /terms, /privacy/, /faq, / (homepage)
3. Styling looks fine with both scss and emotion.js
4. Pages now have page title (e.g., `Aspen | Samples`, `Aspen | Phylogenetic Trees`
5. Tests pass
6. HSTS and CSP work
7. Remove dependencies no longer needed
8. Use `HAPPY_COMMIT` for consistent build id in `next.config.js`. This is needed because FE servers run in a cluster, while each Next.js build by default has a random id, so if Server A sends the browser some static files fingerprinted with its unique build id, and then the browser makes a request for another static file, and such request gets routed to Server B, Server B won't  recognize Server A's build id and thus not being able to find the static asset the browser needs. [Source](https://nextjs.org/docs/api-reference/next.config.js/configuring-the-build-id)
9. Replace `<ProtectedRoute />` component with `useProtectedRoute` hook for AgreeTerms page and Sample page
10. Move font import to root html page with preconnect, so the browser is able to download the font faster!
